SIPS/sip-23.md

```diff
@@ -0,0 +1,599 @@
+---
+sip: 23
+title: JSX for Snap interfaces
+status: Draft
+discussions-to: https://github.com/MetaMask/SIPs/discussions/137
+author: Maarten Zuidhoorn (@mrtenz)
+created: 2024-03-22
+---
+
+## Abstract
+
+This proposal presents a specification for the use of JSX (JavaScript XML) in
+Snap interfaces. It includes a set of definitions, components, and rules for
+usage. The document delves into the details of various components such as
+`SnapElement`, `SnapComponent`, and `SnapNode`. It also elaborates on specific
+elements like , `Address`, `Bold`, `Button`, and `Text` among others, explaining
+their structure and purpose.
+
+Moreover, the proposal takes into account backward compatibility considerations.
+It outlines a systematic approach to translate the components from the previous
+SIP-7 format into the newly proposed format.
+
+The proposal aims to create a more robust, flexible, and versatile system for
+Snap interfaces. It strives to enhance the user experience and improve the
+efficiency of the system by offering a structured and standardised component
+framework.
+
+## Motivation
+
+The motivation behind this proposal is to leverage JSX, a popular syntax
+extension for JavaScript, for designing and implementing Snap interfaces. JSX
+offers several advantages that make it a preferred choice among developers.
+Primarily, it allows for writing HTML-like syntax directly in the JavaScript
+code, which makes it more readable and intuitive. This facilitates easier
+development and maintenance of complex UI structures.
+
+Furthermore, JSX is universally recognised and widely adopted in the JavaScript
+community, especially within the React ecosystem. By using JSX for Snap
+interfaces, we enable a vast number of developers familiar with this syntax to
+contribute effectively in a shorter time frame.
+
+Adopting JSX also ensures better integration with modern development tools and
+practices. It allows for integration with linters, formatters, and type
+checkers, thus improving the development workflow.
+
+In summary, the use of JSX in Snap interfaces aims to improve developer
+experience, enhance code maintainability, and ultimately, lead to the creation
+of more robust and efficient Snap interfaces.
+
+## Specification
+
+> Formal specifications are written in TypeScript.
+
+### Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" written in
+uppercase in this document are to be interpreted as described in
+[RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
```

Comment:
```suggestion
The key words "MUST", "SHOULD", and "MAY" written in
uppercase in this document are to be interpreted as described in
[RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
```
These are the only keywords used in the spec

Reply:
We list all of them in most/all(?) SIPs though

Reply:
Technically, it's an incorrect statement. I get what you're saying though. I'll leave it as a nit then.